pacman/operations/routing_info_allocator_algorithms/zoned_routing_info_allocator.py

```diff
@@ -16,154 +16,308 @@
+        for key_and_mask in constraint.keys_and_masks:
+            # Go through the mask sets and save keys and n_keys
+            for key, n_keys in get_key_ranges(
+                    key_and_mask.key, key_and_mask.mask):
+                self.__fixed_keys.append((key, n_keys))
+
+    def __calculate_zones(self):
         """
-        :return: tuple containing max app keys and a map of app to max keys
-            for any given machine vertex
-        :rtype: tuple(int, dict(ApplicationVertex, int))
         :raises PacmanRouteInfoAllocationException:
         """
+        by_app_and_partition_name = \
+            self.__machine_graph.multicast_partitions
         progress = ProgressBar(
-            self.__application_graph.n_vertices, "Calculating zones")
-        # holders
-        max_app_keys_bits = 0
-        source_zones = 0
-        max_partitions = 0
-        key_bits_per_app = dict()
+            len(by_app_and_partition_name), "Calculating zones")
 
         # search for size of regions
-        for app_vertex in progress.over(self.__application_graph.vertices):
-            app_max_partitions = 0
-            max_keys = 0
-            for vertex in app_vertex.machine_vertices:
-                partitions = self.__machine_graph.\
-                    get_outgoing_edge_partitions_starting_at_vertex(vertex)
-                app_max_partitions = max(app_max_partitions, len(partitions))
-                max_keys = max(max_keys, max(
-                    (self.__n_keys_map.n_keys_for_partition(partition)
-                     for partition in partitions
-                     if partition.traffic_type == EdgeTrafficType.MULTICAST),
-                    default=0))
-            if max_keys > 0:
-                max_partitions = max(max_partitions, app_max_partitions)
-                source_zones += app_max_partitions
-                key_bits = self.__bits_needed(max_keys)
-                machine_bits = self.__bits_needed(len(
-                    app_vertex.machine_vertices))
-                max_app_keys_bits = max(
-                    max_app_keys_bits, machine_bits + key_bits)
-                key_bits_per_app[app_vertex] = key_bits
-        source_bits = self.__bits_needed(source_zones)
-
-        if source_bits + max_app_keys_bits > BITS_IN_KEY:
+        for app_id in progress.over(by_app_and_partition_name):
+            by_app = by_app_and_partition_name[app_id]
+            for partition_name, by_partition_name in by_app.items():
+                max_keys = 0
+                for mac_vertex in by_partition_name:
+                    partition = self.__machine_graph.\
+                        get_outgoing_edge_partition_starting_at_vertex(
+                            mac_vertex, partition_name)
+                    n_keys = self.__n_keys_map.n_keys_for_partition(partition)
+                    max_keys = max(max_keys, n_keys)
+                if max_keys > 0:
```

Comment:
Does this also need to filter out the fixed_keys? As otherwise they will get in the way of the calculations?